drivers/serial/serial.c

```diff
@@ -154,6 +159,28 @@ static struct work_s g_serial_work;
  * Private Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: uart_poll_notify
+ ****************************************************************************/
+
+static void uart_poll_notify(FAR uart_dev_t *dev, unsigned int min,
+                             unsigned int max, pollevent_t eventset)
+{
+  irqstate_t flags;
+
+  DEBUGASSERT(max > min && max - min <= CONFIG_SERIAL_NPOLLWAITERS);
+
+  flags = enter_critical_section();
```

Comment:
I don't think that this fix is correct;

The race condition is inside poll_notify(), when setting the revents bits:
https://github.com/apache/nuttx/blob/904b6ff85b479abe6c75ae2e3e1fda3af9e7a84f/fs/vfs/fs_poll.c#L289

While the poll_notify is being called both from irq and from task context, the read-modify-write needs to be protected there with a critical section in there. This fixes just a symptom for uart driver - the underlying issue is still there in the poll interface and affects other driver using the poll interface.

The sched lock should not be needed while the call to the callback in poll_notify is left out of the critical section - unless you need to protect that the fds->revents doesn't get new bits during a call to callback. But I don't really see the need for that ?

Reply:
sure, let's continue discuss in your new pr.